src/libeep/python/pyeep.c

```diff
@@ -174,6 +202,36 @@ pyeep_get_samples(PyObject* self, PyObject* args) {
 ///////////////////////////////////////////////////////////////////////////////
 static
 PyObject *
+pyeep_get_samples_as_nparray(PyObject* self, PyObject* args) {
+  int handle;
+  int fro;
+  int to;
+  int nd = 2;
+  npy_intp dims[2];
+
+  Py_ssize_t i;
+
+  if(!PyArg_ParseTuple(args, "iii", & handle, & fro, & to)) {
+    return NULL;
+  }
+
+  float * libeep_sample_data = libeep_get_samples(handle, fro, to);
+  if(libeep_sample_data == NULL) {
+    return NULL;
+  }
+
+  dims[0] = to - fro;
+  dims[1] = libeep_get_channel_count(handle);
+  PyObject * numpy_array = PyArray_SimpleNewFromData(nd, dims, NPY_FLOAT32, libeep_sample_data);
+  if(!numpy_array) {
+    return NULL;
+  }
+  PyObject * numpy_array_transposed = PyArray_Transpose(numpy_array, NULL);
+  return numpy_array_transposed;
```

Comment:
Is it 100% required for efficiency at this point to do this in C? I think having a C dependency on NumPy might make it so you can't use abi3 anymore (not 100% sure, though)

So if you have profiled and found that just using `list` here then converting to `ndarray` has a noticable performance impact then it might be worth doing this in C. But if you haven't profiled it might be worth it, keeping/doing this as a `list` and then casting to `np.array` in Python (rather than in C) could be better from a maintenance perspective

Reply:
Arg! Good point, I think you are right and it won't be `abi3` anymore.
On windows, on a ~1h 128 channel file:
- list: 43.7 s ± 940 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
- numpy: 32.5 s ± 36.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

This is an extreme case with a long file, I don't think it's worth the maintenance trouble, it's the same order of magnitude..

Reply:
You could return the `libeep_sample_data` buffer directly and `np.frombuffer` it in Pythotn, in theory that should be very similar to what the C code is actually doing and avoids the need for `list` at all.

Not 100% sure how to handle references and the like though but it might be worth trying to see if it is close to the C speed

Reply:
>> having a C dependency on NumPy might make it so you can't use abi3 anymore.
This is probably not true. Also, the do's and don'ts are well documented here. https://numpy.org/doc/stable/dev/depending_on_numpy.html and a lot of package does rely on numpy-C APIs, so there will be constant support in this regard. @mscheltienne any comments?
 
>> So if you have profiled and found that just using list here then converting to ndarray has a noticable performance impact then it might be worth doing this in C. 
I did do profile before deciding on this. Using a `list`, then converting it to `ndarray` is ~30 times slower than creating the numpy array in C for a small (128x100) array. For longer data segments, it's gonna be even more.

Reply:
> Arg! Good point, I think you are right and it won't be `abi3` anymore. On windows, on a ~1h 128 channel file:
> 
>     * list (`antio` 0.2.0): 43.7 s ± 940 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 
>     * numpy (`antio` 0.3.0.dev0 in this PR): 32.5 s ± 36.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 
> 
> This is an extreme case with a long file, I don't think it's worth the maintenance trouble, it's the same order of magnitude..

What exactly are you profiling? `cnt.get_samples()` vs `cnt.get_samples_as_nparray()` or `mne.io.read_raw_ant()` function calls? I find `cnt.get_samples_as_nparray()` to be ~25-30 times faster than `np.asanyaary(cnt.get_samples())`.

Reply:
> I find cnt.get_samples_as_nparray() to be ~25-30 times faster than np.asanyaary(cnt.get_samples()).

It's going to depend probably on how much data you load at once. If you use a tiny array the Python call time overhead gets worse, so I'd expect your "small matrix" example to have a much worse overhead than what @mscheltienne used. So 25-30x slower for the tiny array vs < 2x slowdown for the "all data" test that @mscheltienne used doesn't sound unreasonable.

But hopefully this point is moot anyway if a `np.frombuffer` approach is used, I'd expect this to be very similar to the C code approach (at least for the "all data" loading... for tiny matrices there will still be some Python call overhead but I think that's okay).

Reply:
Thanks for the idea @larsoner! Took a while to figure out how to return c-array as a python readable buffer, memoryview was the solution. The build is now numpy-free!

@mscheltienne I am not sure if we need the compat-test anymore. I kept it for you to take of it! 
It should be ready to merge after your review.

Reply:
Awesome! 🚀

Reply:
... and @mscheltienne hopefully whatever benchmark you used before shows it's (nearly) as fast as the previous C approach for the whole-data-read case! 🤞

Reply:
I was comparing `cnt. get_samples` and it's `numpy` version. The from buffer version was also my next idea, it's what is used in `mne-lsl`, it's awesome that you worked it out during my night 😂

I'll have a look today at the PR!

Reply:
And performance similar to C: 

```
33.1 s ± 180 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
FYI in case anyone was curious, I was wrong -- you *can* use NumPy in ABI3 mode:

https://github.com/numpy/numpy/issues/27979
https://github.com/numpy/numpy/pull/27981

So if anything gets simpler/easier by reincluding NumPy, we can!